k-distribution/k-tutorial/1_basic/02_basics/README.md

```diff
@@ -1 +1,212 @@
+
+```
+colorOf(Banana())
+```
+
+You can call it whatever file name you want, but for example purposes we will
+call this file `banana.color`. Save it in the same directory that also contains
+the `lesson-02-a-kompiled` directory.
+
+We can now evaluate this K term by running:
+
+```
+krun banana.color
+```
+
+From that directory. `krun` will invoke K's interpreter by interpreting a
+program according to the definition that was compiled in that directory.
+
+You will get the following output:
+
+```
+<k>
+  Yellow ( ) ~> .
+</k>
+```
+
+For now, don't worry about the `<k>`, `</k>`, or ` ~> .` portions of this
+output file. Instead, let's look at what this definition and program did.
+
+## Productions, Constructors, and Functions
+
+The first thing to realize is that this K definition contains 5 productions.
+Productions are introduced with the **syntax** keyword, followed by a **sort**,
+followed by the operator `::=`, followed by one or more **constructors**,
+**functions**, **tokens**, **brackets**, **lists**, **macros**, **aliases**, 
+or **anywhere productions**, separated by the `|` operator. Each of these
+individually is a single production, and the `|` symbol simply groups together
+productions that have the same sort. For example, we could equally have written
+an identical K definition like so:
+
+```k
+module LESSON-02-B
+
+  syntax Color ::= Yellow()
+  syntax Color ::= Blue()
+  syntax Fruit ::= Banana()
+  syntax Fruit ::= Blueberry()
+  syntax Color ::= colorOf(Fruit) [function]
+
+  rule colorOf(Banana()) => Yellow()
+  rule colorOf(Blueberry()) => Blue()
+
+endmodule
+```
+
+Or even:
+
+```k
+module LESSON-02-C
+
+  syntax Color ::= Yellow()
+                 | Blue()
+                 | colorOf(Fruit) [function]
+  syntax Fruit ::= Banana()
+                 | Blueberry()
+
+  rule colorOf(Banana()) => Yellow()
+  rule colorOf(Blueberry()) => Blue()
+
+endmodule
+```
+
+Each of these types of productions named above has the same underlying syntax,
+but context and **attributes** are used to distinguish between the different
+types. Tokens, brackets, lists, macros, aliases, and anywhere productions will
+be covered in a later lesson, but this lesson does introduce us to constructors
+and functions. `Yellow()`, `Blue()`, `Banana()`, and `Blueberry()` are
+constructors. You can think of a constructor like a constructor for an
+algebraic data type, if you're familiar with a functional language. The data
+type itself is the sort that appears on the left of the `::=` operator. Sorts
+in K consist of uppercase identifiers.
+
+Constructors can have arguments, but these constructors do not have any
+arguments. We will cover the syntax of productions in detail in the next
+lesson, but for now, you can write a production with no arguments as an
+uppercase or lowercase identifier followed by the `()` operator. You can write
+a production with arguments by putting the sorts of the arguments in a
+comma-separated list inside the parentheses.
```

Comment:
```suggestion
uppercase or lowercase identifier followed by the `()` operator.
```

Don't explain things that are not introduced yet. In fact, I wouldn't even mention that you can have `alias`, `macro`, etc... other types of declarations, you can just say "these are _some_ of the types of productions you can define".

The point being that anything explained that _is not immediately used_ is just confusing to the user. The basic rules of writing lessons are (i) don't explain anything that is not immediately useful for the lesson, and (ii) make sure you do explain _everything_ that is needed for the lesson.